usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,194 @@
+)
+
+// Parser takes care of unmarshaling a config struct
+// from given raw bytes(e.g: YAML, JSON, etc).
+type Parser[T any] func([]byte) (*T, error)
+
+// ConfigManager takes care of periodically loading the config from
+// given filepath for every interval period.
+type ConfigManager[T any] struct {
+	// path is file path of config to load and unmarshal from
+	path string
+	// interval is how often config manager trigger loading the config file.
+	interval time.Duration
+	// parse takes care of unmarshaling the config struct from a file
+	parse Parser[T]
+
+	// currentConfig is last successfully loaded config.
+	// ConfigManager keep using this config if there are any
+	// failures to load new configs.
+	currentConfig *T
+	currentHash   string
+	mu            sync.RWMutex // protects currentConfig
+
+	log             logrus.FieldLogger
+	lastLoadSuccess prometheus.Gauge
+	configHash      *prometheus.GaugeVec
+}
+
+func NewConfigManager[T any](
+	filepath string,
+	parser Parser[T],
+	interval time.Duration,
+	log logrus.FieldLogger,
+	r prometheus.Registerer,
+) (*ConfigManager[T], error) {
+	// catch empty filepath early
+	if len(strings.TrimSpace(filepath)) == 0 {
+		return nil, errors.New("filepath to load runtimeconfig is empty")
+	}
+
+	cm := &ConfigManager[T]{
+		path:     filepath,
+		parse:    parser,
+		interval: interval,
+		log:      log,
+		lastLoadSuccess: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_last_load_success",
+			Help: "Whether the last loading attempt of runtime config was success",
+		}),
+		configHash: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_hash",
+			Help: "Hash value of the currently active runtime configuration",
+		}, []string{"sha256"}), // sha256 is type of checksum and hard-coded for now
+	}
+
+	// try to load it once to fail early if configs are invalid
+	if err := cm.loadConfig(); err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+// Run is a blocking call that starts the configmanager actor. Consumer probably want to
+// call it in different groutine. It also respects the passed in `ctx`.
+// Meaning, cancelling the passed `ctx` stops the actor.
+func (cm *ConfigManager[T]) Run(ctx context.Context) error {
+	return cm.loop(ctx)
+	// cm.log.Info("starting runtime config manager")
+
+	// // Cancelling the passed in context is graceful shutdown
+	// if err := cm.loop(ctx); !errors.Is(err, context.Canceled) {
+	// 	cm.log.Error("stopping runtime config manager due to error", "msg", err.Error())
+	// }
+
+	// cm.log.Info("stopping runtime config manager")
+	// return nil
+}
+
+// Config returns the current valid config if available. Once the config manager
+// is started without any error, consumer should be able to get **valid** config
+// via this api.
+func (cm *ConfigManager[T]) Config() (*T, error) {
+	cm.mu.RLock()
+	defer cm.mu.RUnlock()
+
+	if cm.currentConfig == nil {
+		return nil, ErrEmptyConfig
+	}
```

Comment:
IIUC, the config should be never `nil` because you do a mandatory load in the constructor

Reply:
Sure. not at the point. It's just an extra safety check to have an invariant of non-nil config at this point in code. To catch any future bugs when adding features to it.